### PR TITLE
Reset `self.best` in EarlyStopping/ReduceLROnPlateau between fit calls

### LINKED ISSUE #20256: Early-stopping does not work properly in Keras 3 when used in a for loop
Hello, 
I am using Keras 3.5 with TF 2.17. My code is more or less the following (but it is not a grid search as in the real code I also increment some other variables that are not directly linked to the network):
```

def create_conv(nb_cc_value, l1_value):
    model = Sequential()
    model.add(tensorflow.keras.layers.RandomFlip(mode="horizontal"))
    model.add(Conv2D(32, (3,3), activation = 'relu', kernel_regularizer=l1(l1_value)))
    model.add(MaxPool2D())
    model.add(BatchNormalization())
    model.add(Conv2D(64, (3,3), activation = 'relu', kernel_regularizer=l1(l1_value)))
    model.add(MaxPool2D())
    model.add(BatchNormalization())
    model.add(Conv2D(512, (3,3), activation = 'relu', kernel_regularizer=l1(l1_value)))
    model.add(MaxPool2D())
    model.add(BatchNormalization())
    model.add(Conv2D(1024, (3,3), activation = 'relu', kernel_regularizer=l1(l1_value)))
    model.add(BatchNormalization())
    model.add(MaxPool2D())
    model.add(Conv2D(2048, (3,3), activation = 'relu', kernel_regularizer=l1(l1_value)))
    model.add(Flatten())
    model.add(BatchNormalization())
    model.add(Dense(nb_cc_value, activation='relu', kernel_regularizer=l1(l1_value)))
    model.add(Dense(56, activation = 'sigmoid'))
    model.build((None,150,150,1))
    
    lr_schedule = tensorflow.keras.optimizers.schedules.ExponentialDecay(initial_learning_rate=0.01, decay_steps=10000, decay_rate=0.7, staircase=False)
    optimizer = tensorflow.keras.optimizers.SGD(learning_rate=lr_schedule, momentum = 0.9)
    model.compile(loss= ['mse'], optimizer = optimizer, metrics = ['mse'])
    return model

# %%--------------------------------------------------Initialization
early_stopping = EarlyStopping(monitor='val_mse', min_delta = 0.001, patience=5, restore_best_weights=True)

nb_cc = [2, 6, 12, 102, 302, 602]
l1_values = [2.220446049250313e-16, 0.0000001, 0.0001]

for nb_cc_value in nb_cc:
    for l1_value in l1_values:
        for run in range(1,3):
            model = create_conv(nb_cc_value, l1_value)
            history = model.fit(X_train, y_train, epochs=epoques,callbacks=[early_stopping], validation_data=(X_test, y_test), batch_size=6, shuffle=True, verbose=1)
                # Nettoyage
            del X_train, y_train, X_test, y_test, vectors_dict, ethnie_dict, test_image_counts, model, history, prediction
            tensorflow.keras.backend.clear_session()
            gc.collect()

```

However, when I run it, only the very first run in the whole code works fine. The others all stops at something like 1 or 2 epochs even if the 'val_mse' variable is decreasing. I have run it using Keras 2.15.0 (tensorflow 2.15.0.post1) and it worked fine then. 

Any help is much appreciated, thank you

### PR DESCRIPTION
**Pull request by rstar327**
## Description

Fixes #20256.

`EarlyStopping` and `ReduceLROnPlateau` store the best-seen monitored metric in `self.best`. Both already reset `self.wait`, `self.cooldown_counter`, etc. in `on_train_begin` so they can be reused across `model.fit()` calls — but they forget to reset `self.best`. When the same callback instance is passed to multiple `fit()` calls (a common pattern when sweeping hyperparameters), the second call's first epoch is compared against the *previous* model's best metric, which is meaningless, and training stops prematurely.

### Reproduction

```python
import keras, numpy as np
from keras import callbacks, layers, models

data = np.random.random((100, 1)).astype("float32")
labels = np.where(data > 0.5, 1, 0).astype("float32")
stopper = callbacks.EarlyStopping(monitor="loss", patience=5)

for i in range(3):
    model = models.Sequential([
        layers.Dense(10, activation="relu"),
        layers.Dense(1, activation="sigmoid"),
    ])
    model.compile(optimizer="sgd", loss="binary_crossentropy")
    history = model.fit(data, labels, callbacks=[stopper], verbose=0, epochs=50)
    print(f"Run {i}: {len(history.epoch)} epochs")
```

**Before:** `Run 0: 50, Run 1: 5 (stale best), Run 2: 50`
**After:** `Run 0: 50, Run 1: 50, Run 2: 50`

### Implementation

- `keras/src/callbacks/early_stopping.py:on_train_begin`: clear `self.best`. The base `_is_improvement(monitor_value, None)` already returns `True`, so the first epoch becomes the new best naturally.
- `keras/src/callbacks/reduce_lr_on_plateau.py:_reset`: same — clear `self.best`. (`_reset` is already called from `on_train_begin`.)

Added regression tests:
- `EarlyStoppingTest::test_early_stopping_reuse_across_models` — three sequential `model.fit()` calls each get a fresh model and must each run more than `patience` epochs.
- `ReduceLROnPlateauTest::test_best_reset_across_fit_calls` — direct unit test that `self.best` is `None` after a fresh `on_train_begin`.

The fix is the same idea as the one in the abandoned PR #22358, which was closed by the stale bot. Picking it up since the bug still reproduces on master.

## Contributor Agreement

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

### CODE REVIEW

File: keras/src/layers/attention/multi_head_attention.py [hunk 1]
```diff
@@ -134,6 +134,24 @@ def __init__(
...
+                f"a positive integer. Received: key_dim={key_dim}"
+            )
+        if value_dim is not None and (
+            not isinstance(value_dim, int) or value_dim <= 0
+        ):
+            raise ValueError(
+                "Received an invalid value for argument `value_dim`, "
+                "expected a positive integer or `None`. Received: "
+                f"value_dim={value_dim}"
+            )
```
Comment (hertschuh) [inline keras/src/layers/attention/multi_head_attention.py:154]: Undo this file, this was another PR

File: keras/src/layers/normalization/group_normalization.py [hunk 2]
```diff
@@ -83,6 +83,12 @@ def __init__(
         gamma_constraint=None,
         **kwargs,
     ):
+        if not isinstance(groups, int) or (groups <= 0 and groups != -1):
+            raise ValueError(
+                "Received an invalid value for argument `groups`, expected "
+                "a positive integer (or `-1` for instance normalization). "
+                f"Received: groups={groups}"
+            )
```
Comment (hertschuh) [inline keras/src/layers/normalization/group_normalization.py:91]: Undo this file, this was another PR

File: keras/src/layers/normalization/group_normalization_test.py [hunk 3]
```diff
@@ -72,6 +72,13 @@ def test_groups_not_a_multiple_of_dim_error(self):
         ):
             _ = layer(inputs)
 
+    def test_groups_rejects_invalid_values(self):
+        for bad in (0, -2, 1.5, "32"):
+            with self.assertRaisesRegex(ValueError, "argument `groups`"):
+                layers.GroupNormalization(groups=bad)
+        # `-1` is the documented sentinel for instance normalization.
+        layers.GroupNormalization(groups=-1)
```
Comment (hertschuh) [inline keras/src/layers/normalization/group_normalization_test.py:80]: Undo this file, this was another PR

File: keras/src/layers/attention/multi_head_attention_test.py [hunk 4]
```diff
@@ -394,6 +394,17 @@ def test_shape_mismatch_error(self, query_shape, value_shape, key_shape):
...
+        for bad in (0, -2, 1.5):
+            with self.assertRaisesRegex(ValueError, "argument `num_heads`"):
+                layers.MultiHeadAttention(num_heads=bad, key_dim=4)
+            with self.assertRaisesRegex(ValueError, "argument `key_dim`"):
+                layers.MultiHeadAttention(num_heads=4, key_dim=bad)
+            with self.assertRaisesRegex(ValueError, "argument `value_dim`"):
+                layers.MultiHeadAttention(num_heads=4, key_dim=4, value_dim=bad)
+        # `value_dim=None` is the documented default (mirror `key_dim`).
+        layers.MultiHeadAttention(num_heads=4, key_dim=4, value_dim=None)
+
```
Comment (hertschuh) [inline keras/src/layers/attention/multi_head_attention_test.py:407]: Undo this file, this was another PR

File: keras/src/callbacks/reduce_lr_on_plateau_test.py [hunk 5]
```diff
@@ -139,3 +139,22 @@ def test_cooldown(self):
...
+        reduce_lr = callbacks.ReduceLROnPlateau(
+            patience=1, factor=0.1, monitor="loss", min_delta=0
+        )
+        reduce_lr.set_model(self.model)
+        reduce_lr.on_train_begin()
+        reduce_lr.on_epoch_end(0, logs={"loss": 0.5})
+        self.assertEqual(reduce_lr.best, 0.5)
+        # New `model.fit()` -> `on_train_begin` should clear `self.best`.
+        reduce_lr.on_train_begin()
+        self.assertIsNone(reduce_lr.best)
```
Comment (hertschuh) [inline keras/src/callbacks/reduce_lr_on_plateau_test.py:159]: Verify the other fields too: `cooldown_counter`, `wait`.

File: keras/src/callbacks/early_stopping_test.py [hunk 6]
```diff
@@ -126,6 +126,30 @@ def test_early_stopping_reuse(self):
...
+                [
+                    layers.Dense(10, activation="relu"),
+                    layers.Dense(1, activation="sigmoid"),
+                ]
+            )
+            model.compile(optimizer="sgd", loss="binary_crossentropy")
+            history = model.fit(
+                data, labels, callbacks=[stopper], verbose=0, epochs=20
+            )
+            self.assertGreater(len(history.epoch), patience)
```
Comment (hertschuh) [inline keras/src/callbacks/early_stopping_test.py]: It's hard for me to tell whether this will always work considering that the data is random.

So I think I would prefer the same approach as what you did with `keras/src/callbacks/reduce_lr_on_plateau_test.py`, which will also be much faster to execute.

**Review by hertschuh — CHANGES_REQUESTED** [review]
Thanks for fixing this!

**Comment by rstar327** [comment]
Thanks for the review! Updated:

1. **Dropped the GroupNorm / MultiHeadAttention changes** — those belong to #22966, they got pulled into this branch by mistake. The PR now only touches the two callback files and their tests. Verified via `git diff upstream/master --stat`.

2. **Replaced `test_early_stopping_reuse_across_models`** with a direct, deterministic unit-style test (`test_state_reset_across_fit_calls`) that mirrors the ReduceLROnPlateau test. It manually sets per-run state on the callback, calls `on_train_begin` again, and asserts every field is reset (`best`, `wait`, `stopped_epoch`, `best_epoch`, `best_weights`). No random data, no `model.fit`, runs in milliseconds.

3. **Extended the `ReduceLROnPlateau` test** to verify `cooldown_counter` and `wait` are reset alongside `self.best`.

**Review by hertschuh — APPROVED** [review]
Thanks!
